python/ray/data/datasource/datasource.py

```diff
@@ -100,6 +99,50 @@ def on_write_failed(
         pass
 
 
+@DeveloperAPI
+class Reader(Generic[T]):
```

Comment:
I'm concerned that introducing this `Reader` concept in addition to `Datasource` will make implementing custom datasources more difficult/confusing. The common case of subclassing `FileBasedDatasource` is covered since such subclasses can be `Reader`-agnostic, which is good, so this slight custom datasource UX degradation doesn't seem like a blocker.

It seems like the core reason that we can't do this read-time binding on the `Datasource` itself via its constructor (consolidating `Datasource` and `Reader`) is that the binding will need to be able to happen in a remote task, although we could work around this if we had `read_datasource` take an uninstantiated `Datasource`.

What do you think of this alternative, where the `Datasource` constructor is treated as the read-time binding? Note that this would be a breaking change for existing custom datasources.

```python
# datasource.py

class Datasource(Generic[T]):
    def estimate_inmemory_data_size(self) -> Optional[int]:
        return None

    def prepare_read(self, parallelism: int) -> List["ReadTask[T]"]:
        raise NotImplementedError

# file_based_datasource.py

class FileBasedDatasource(Datasource):
    def __init__(
        self,
        paths: Union[str, List[str]],
        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
        schema: Optional[Union[type, "pyarrow.lib.Schema"]] = None,
        open_stream_args: Optional[Dict[str, Any]] = None,
        meta_provider: BaseFileMetadataProvider = DefaultFileMetadataProvider(),
        partition_filter: PathPartitionFilter = None,
        # TODO(ekl) deprecate this once read fusion is available.
        _block_udf: Optional[Callable[[Block], Block]] = None,
        **reader_args,
    ):
        _check_pyarrow_version()
        self._schema = schema
        self._open_stream_args = open_stream_args
        self._meta_provider = meta_provider
        self._partition_filter = partition_filter
        self._block_udf = _block_udf
        self._reader_args = reader_args
        paths, self._filesystem = _resolve_paths_and_filesystem(paths, filesystem)
        self._paths, self._file_sizes = meta_provider.expand_paths(
            paths, self._filesystem
        )

    def estimate_inmemory_data_size(self) -> Optional[int]:
        total_size = 0
        for sz in self._file_sizes:
            if sz is not None:
                total_size += sz
        return total_size

    def prepare_read(self, parallelism: int) -> List[ReadTask]:
        # ...

# read_api.py

@PublicAPI
def read_csv(
    paths: Union[str, List[str]],
    *,
    filesystem: Optional["pyarrow.fs.FileSystem"] = None,
    parallelism: int = 200,
    ray_remote_args: Dict[str, Any] = None,
    arrow_open_stream_args: Optional[Dict[str, Any]] = None,
    meta_provider: BaseFileMetadataProvider = DefaultFileMetadataProvider(),
    partition_filter: PathPartitionFilter = None,
    **arrow_csv_args,
) -> Dataset[ArrowRow]:
    return read_datasource(
        CSVDatasource,
        parallelism=parallelism,
        paths=paths,
        filesystem=filesystem,
        ray_remote_args=ray_remote_args,
        open_stream_args=arrow_open_stream_args,
        meta_provider=meta_provider,
        partition_filter=partition_filter,
        **arrow_csv_args,
    )

def read_datasource(
    datasource: Type[Datasource[T]],
    *,
    parallelism: int = -1,
    ray_remote_args: Dict[str, Any] = None,
    **read_args,
) -> Dataset[T]:
    # ...

    if force_local:
        read_tasks = _prepare_read(datasource, ctx, cur_pg, parallelism, read_args)
    else:
        # Prepare read in a remote task so that in Ray client mode, we aren't
        # attempting metadata resolution from the client machine.
        prepare_read = cached_remote_fn(
            _prepare_read, retry_exceptions=False, num_cpus=0
        )
        read_tasks = ray.get(
            prepare_read.remote(
                datasource,
                ctx,
                cur_pg,
                parallelism,
                read_args,
            )
        )

    # ...

def _prepare_read(
    ds_cls: Type[Datasource],
    ctx: DatasetContext,
    cur_pg: Optional[PlacementGroup],
    parallelism: int,
    kwargs: Dict[str, Any],
) -> List[ReadTask]:
    kwargs = _unwrap_arrow_serialization_workaround(kwargs)
    DatasetContext._set_current(ctx)
    ds = ds_cls(**kwargs)
    parallelism = _autodetect_parallelism(parallelism, cur_pg, ds)
    return ds.prepare_read(parallelism)
```

Reply:
I thought about that, but I think the current approach is a little better in that:
1. We retain backwards compatibility, and
2. You get the flexibility of a single datasource class that can be used for multiple operations.

Probably either could work, but erring on the side of not breaking APIs.

Reply:
>You get the flexibility of a single datasource class that can be used for multiple operations.

Hmm but all existing datasources are constructed and used on a per-operation basis, and I can't immediately think of why we would want to reuse a datasource instance across multiple operations. That only makes sense if there's job/session-level data that the user wants to only supply once, and then apply to many read operations... It seems like this requirement might be a case of YAGNI.

> We retain backwards compatibility

I think that ensuring backwards compatibility is probably a good enough reason to keep the current `Reader` implementation. We'll just need to make sure that the custom datasource docs thoroughly explain `Datasource` vs `Reader`, and how to implement both.

Reply:
Since we are prior GA, maybe breaking APIs is less a concern (than e.g. changing ray.remote) and will be worth it to get it right?

Reply:
Personally this pattern seems very ordinary / normal to me, so I'd push to keep it and also retain backwards compatibility. We should not be in the habit of breaking APIs unless there are very good reasons.

Reply:
Sounds good, after re-read it seems not a concern to me.